pkg/controller/node/ipam/cidr_allocator.go

```diff
@@ -37,15 +48,72 @@ const (
 	// CloudAllocatorType is the allocator that uses cloud platform
 	// support to do node CIDR range allocations.
 	CloudAllocatorType CIDRAllocatorType = "CloudAllocator"
+	// IPAMFromClusterAllocatorType uses the ipam controller sync'ing the node
+	// CIDR range allocations from the cluster to the cloud.
+	IPAMFromClusterAllocatorType = "IPAMFromCluster"
+	// IPAMFromCloudAllocatorType uses the ipam controller sync'ing the node
+	// CIDR range allocations from the cloud to the cluster.
+	IPAMFromCloudAllocatorType = "IPAMFromCloud"
+
+	// The amount of time the nodecontroller polls on the list nodes endpoint.
+	apiserverStartupGracePeriod = 10 * time.Minute
 )
 
-// CIDRAllocator is an interface implemented by things that know how to
-// allocate/occupy/recycle CIDR for nodes.
+// CIDRAllocator is an interface implemented by things that know how
+// to allocate/occupy/recycle CIDR for nodes.
 type CIDRAllocator interface {
 	// AllocateOrOccupyCIDR looks at the given node, assigns it a valid
 	// CIDR if it doesn't currently have one or mark the CIDR as used if
 	// the node already have one.
 	AllocateOrOccupyCIDR(node *v1.Node) error
 	// ReleaseCIDR releases the CIDR of the removed node
 	ReleaseCIDR(node *v1.Node) error
+	// Register allocator with the nodeInformer for updates.
+	Register(nodeInformer informers.NodeInformer)
+}
+
+// New creates a new CIDR range allocator.
+func New(kubeClient clientset.Interface, cloud cloudprovider.Interface, allocatorType CIDRAllocatorType, clusterCIDR, serviceCIDR *net.IPNet, nodeCIDRMaskSize int) (CIDRAllocator, error) {
+	nodeList, err := listNodes(kubeClient)
+	if err != nil {
+		return nil, err
+	}
+
+	var cidrAllocator CIDRAllocator
+
+	switch allocatorType {
+	case RangeAllocatorType:
```

Comment:
I think you can consolidate this a little since  you return the allocator and error 

remove the temp var cidrAllocator

and just have 

switch allocatorType {
 	case RangeAllocatorType:
         return NewCloudCIDRAllocator(kubeClient, cloud)
    case  CloudAllocatorType:
         return NewCloudCIDRAllocator(kubeClient, cloud)
    default: 
         return nil, fmt.Errorf("...")
}

not need for error check or return at that end.

Reply:
thanks!